### PR TITLE
[FLINK-35702][format][parquet] support parquet read nested.

### PR DESCRIPTION
**Pull request by Stephen0421**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*
This pull request supports parquet read nested data such as array<array>, array<map>, array<row>.

## Brief change log

  -  add new structure to record every column's repetition level and definition level.
  - add utils to support parse correct structure with repetition level and definition level.
  - make HeapRowVector supports set child fields
  - make HeapMapVector supports set key vector and value vector

## Verifying this change

Please make sure both new and modified tests in this PR follow [the conventions for tests defined in our code quality guide](https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#7-testing).

This change added tests and can be verified as follows:

  - Added test case to validate the parsing correct nested structure.
  - Manually verified the change by running flink jobs.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive):no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no 
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  yes
  - If yes, how is the feature documented? JavaDocs
  - 
### About how to use definition level and repetition level to parse data
As we know each column of parquet has three type values: 
- data
- repetition level
- definition level
Only repeated fields increment the repetition level, only non-required fields increment the definition level.
#### Repetition Level
Let’s use the following example to explain how repetition level works.
Here is a message type of parquet:
```
message nested {
  repeated group leve1 {
    repeated string leve2;
  }
}
```

And the column value is as below:
Data:                        a, b, c, d, e, f, g, h, i, j
Repetition leve(R):  0, 2, 2, 1, 2, 2, 2, 0, 1, 2, (0)
In general, R = 0 means this is a new row. So a is the first element of the inner array: [[a]]
For b, the R = 2 means it belongs to level2, so it belongs to the first inner array, so as c : [[a, b, c]]
For d, the R = 1 means it belongs to level1, which means it belongs to a new inner array: [[a, b, c], [d]]
For e, f, g, it’s the same as b and c, so when read to g, the row will be [[a, b, c], [d, e, f, g]].
For h, the R = 0, which means this is the beginning of a new row.
so whole data will be:
R1: [[a, b, c], [d, e, f, g]]
R2: [[h], [i, j]]

#### Definition Level
Definition used to judge whether the value is existing or not.
Here is a message type of parquet:
```
message ExampleDefinitionLevel {
  optional group a {
    optional group b {
      optional string c;
    }
  }
}
```

For this message, it contains a column "a.b.c" where each node of the column is of optional type. When "c" is defined, both "a" and "b" are guaranteed to be defined. When "c" is undefined, we need to indicate from which level it is undefined. The value and meaning is as below:

| Value | Definition Level |
|---------|---------|
| a:null | 0 |
|a:{b:null} | 1 |
|a:{b:{c:null}} | 2 |
|a:{b:{c:”foo”}} |  3(all defined)|

### CODE REVIEW

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+     * @return {@link CollectionPosition} contains collections offset array, length array and isNull
+     *     array.
+     */
+    public static CollectionPosition calculateCollectionOffsets(
+            Field field, int[] definitionLevels, int[] repetitionLevels) {
+        int collectionDefinitionLevel = field.getDefinitionLevel();
+        int collectionRepetitionLevel = field.getRepetitionLevel() + 1;
+        int offset = 0;
+        int valueCount = 0;
+        LongArrayList offsets = new LongArrayList();
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java:103]: Take a look to `org.apache.flink.runtime.util.LongArrayList`.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+     * @param fieldRepetitionLevels int array with each value's repetition level.
+     * @param fieldDefinitionLevels int array with each value's definition level.
+     * @return {@link RowPosition} contains collections row count and isNull array.
+     */
+    public static RowPosition calculateRowOffsets(
+            Field field, int[] fieldDefinitionLevels, int[] fieldRepetitionLevels) {
+        int rowDefinitionLevel = field.getDefinitionLevel();
+        int rowRepetitionLevel = field.getRepetitionLevel();
+        int nullValuesCount = 0;
+        BooleanArrayList nullRowFlags = new BooleanArrayList();
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java:54]: You can add a `BooleanArrayList` just like  `org.apache.flink.runtime.util.LongArrayList`.

File: flink-connectors/flink-connector-hive/pom.xml [hunk 3]
```diff
@@ -1254,6 +1254,7 @@ under the License.
 									<include>org.apache.parquet:parquet-encoding</include>
 									<include>org.apache.parquet:parquet-format-structures</include>
 									<include>org.apache.parquet:parquet-jackson</include>
+									<include>it.unimi.dsi:fastutil</include>
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/pom.xml:1257]: See comments above, we don't need to introduce this dependency.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetDecimalVector.java [hunk 4]
```diff
@@ -69,4 +75,157 @@ public ColumnVector getVector() {
...
+        }
+        return false;
+    }
+
+    @Override
+    public WritableIntVector reserveDictionaryIds(int capacity) {
+        if (vector instanceof WritableColumnVector) {
+            return ((WritableColumnVector) vector).reserveDictionaryIds(capacity);
+        }
+        return null;
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetDecimalVector.java:127]: Can we just throw exception?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/MapColumnReader.java [hunk 5]
```diff
@@ -25,6 +25,7 @@
 import java.io.IOException;
 
 /** Map {@link ColumnReader}. */
+@Deprecated
 public class MapColumnReader implements ColumnReader<WritableColumnVector> {
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/MapColumnReader.java:29]: Why not just delete it?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ArrayColumnReader.java [hunk 6]
```diff
@@ -44,6 +44,7 @@
 import java.util.List;
 
 /** Array {@link ColumnReader}. TODO Currently ARRAY type only support non nested case. */
+@Deprecated
 public class ArrayColumnReader extends BaseVectorizedColumnReader {
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ArrayColumnReader.java:48]: Why not just delete it?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RowColumnReader.java [hunk 7]
```diff
@@ -24,6 +24,7 @@
 import java.util.List;
 
 /** Row {@link ColumnReader}. */
+@Deprecated
 public class RowColumnReader implements ColumnReader<WritableColumnVector> {
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RowColumnReader.java:28]: Why not just delete it?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/Field.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.parquet.vector.type;
+
+import org.apache.flink.table.types.logical.LogicalType;
+
+/** Field that represent parquet's field type. */
+public abstract class Field {
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/Field.java:24]: ParquetField?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/GroupField.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.apache.flink.shaded.guava31.com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+/** Field that represent parquet's Group Field. */
+public class GroupField extends Field {
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/GroupField.java:30]: ParquetGroupField

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/PrimitiveField.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+package org.apache.flink.formats.parquet.vector.type;
+
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.apache.parquet.column.ColumnDescriptor;
+
+import static java.util.Objects.requireNonNull;
+
+/** Field that represent parquet's primitive field. */
+public class PrimitiveField extends Field {
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/PrimitiveField.java:28]: ParquetPrimitiveField

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java (hunk 1 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java]: done

File: flink-connectors/flink-connector-hive/pom.xml (hunk 3 above)
Comment (Stephen0421) [reply flink-connectors/flink-connector-hive/pom.xml]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetDecimalVector.java (hunk 4 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetDecimalVector.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/MapColumnReader.java (hunk 5 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/MapColumnReader.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ArrayColumnReader.java (hunk 6 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ArrayColumnReader.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RowColumnReader.java (hunk 7 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RowColumnReader.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/Field.java (hunk 8 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/Field.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/GroupField.java (hunk 9 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/GroupField.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/PrimitiveField.java (hunk 10 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/type/PrimitiveField.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedColumnReader.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This ColumnReader mainly used to read `Group` type in parquet such as `Map`, `Array`, `Row`. The
+ * method about how to resolve nested struct mainly refer to :
+ * https://github.com/julienledem/redelm/wiki/The-striping-and-assembly-algorithms-from-the-Dremel-paper.
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedColumnReader.java:52]: You can use HTML link

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+            // nested collections/structs
+            if (repetitionLevels[nextIndex] <= maxRepetitionLevel) {
+                size++;
+            }
+            nextIndex++;
+        }
+        return size;
+    }
+
+    private static boolean isCollectionBeginningMarker(
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java:201]: revert this?
isNotCollectionBeginningMarker

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/position/RowPosition.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.parquet.vector.position;
+
+import java.util.Optional;
+
+/** To represent struct's position in repeated type. */
+public class RowPosition {
+    private final Optional<boolean[]> isNull;
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/position/RowPosition.java:25]: ditto

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedPrimitiveColumnReader.java [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+public class NestedPrimitiveColumnReader extends BaseVectorizedColumnReader {
+
+    // flag to indicate if there is no data in parquet data page
+    private boolean eof = false;
+
+    private boolean isFirstRow = true;
+
+    private Object lastValue;
+
+    private final IntArrayList repetetionLevelList = new IntArrayList();
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedPrimitiveColumnReader.java:56]: repetitionLevelList

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java [hunk 15]
```diff
@@ -578,4 +569,167 @@ public static WritableColumnVector createWritableColumnVector(
...
+                    type, repetitionLevel, definitionLevel, required, ImmutableList.of(field));
+        }
+
+        PrimitiveColumnIO primitiveColumnIO = (PrimitiveColumnIO) columnIO;
+        return new ParquetPrimitiveField(
+                type, required, primitiveColumnIO.getColumnDescriptor(), primitiveColumnIO.getId());
+    }
+
+    public static ColumnIO lookupColumnByName(GroupColumnIO groupColumnIO, String columnName) {
+        ColumnIO columnIO = groupColumnIO.getChild(columnName);
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java:686]: Sorry, I think this part i explain wrong just now. When getting columnIO, we firstly use the original name to get the columnIO, and if we can not get it, we will try to iterate every child columnIO and match the column name ignore case.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java [hunk 16]
```diff
@@ -578,4 +569,167 @@ public static WritableColumnVector createWritableColumnVector(
...
+                    required,
+                    ImmutableList.of(keyField, valueField));
+        }
+
+        if (type instanceof ArrayType) {
+            ArrayType arrayType = (ArrayType) type;
+            ColumnIO elementTypeColumnIO;
+            if (columnIO instanceof GroupColumnIO) {
+                GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
+                if (!StringUtils.isNullOrWhitespaceOnly(filedName)) {
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java:651]: When the field name is null, it means the relevant array is double nested array like the inner array of array<array<string>>, so it doesn't have name. In this situation we just need to get the child of the parent ColumnIO.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java:686]: OK, so the following code is only for addressing case insensitive issue

**Comment by empathy87** [comment]
@Stephen0421, could you, please, look at this issue https://issues.apache.org/jira/projects/FLINK/issues/FLINK-35698
It seems to be connected to your work. There is a problem Parquet reading nested data when field type is DECIMAL. I provided a test case there.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java:686]: Yes

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java [hunk 17]
```diff
@@ -578,4 +569,167 @@ public static WritableColumnVector createWritableColumnVector(
...
+                    if (arrayType.getElementType() instanceof RowType) {
+                        elementTypeColumnIO = groupColumnIO;
+                    } else {
+                        elementTypeColumnIO = groupColumnIO.getChild(0);
+                    }
+                }
+            } else if (columnIO instanceof PrimitiveColumnIO) {
+                elementTypeColumnIO = columnIO;
+            } else {
+                throw new RuntimeException(String.format("Unkown ColumnIO, %s", columnIO));
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java]: done

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+                // when definitionLevels[i] < collectionDefinitionLevel - 1, it means the collection
+                // is
+                // not defined, but we need to regard it as null to avoid getting value wrong.
+                nullCollectionFlags.add(true);
+                nullValuesCount++;
+                offsets.add(++offset);
+                emptyCollectionFlags.add(false);
+            }
+        }
+        long[] offestsArray = offsets.toArray();
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java]: ok

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/position/CollectionPosition.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.parquet.vector.position;
+
+import java.util.Optional;
+
+/** To represent collection's position in repeated type. */
+public class CollectionPosition {
+    private final Optional<boolean[]> isNull;
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/position/CollectionPosition.java]: ok

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedPrimitiveColumnReader.java [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+import shaded.parquet.it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Reader to read nested primitive column. */
+public class NestedPrimitiveColumnReader extends BaseVectorizedColumnReader {
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedPrimitiveColumnReader.java]: done

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/vector/ParquetColumnarRowSplitReaderTest.java [hunk 21]
```diff
@@ -592,6 +635,38 @@ void testPartitionValues(int rowGroupSize) throws IOException {
         innerTestPartitionValues(testPath, partSpec, true, rowGroupSize);
     }
 
+    @ParameterizedTest
+    @MethodSource("parameters")
+    public void testNestedRead(int rowGroupSize) throws Exception {
```
[the remark this answers is not in the document]
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/vector/ParquetColumnarRowSplitReaderTest.java]: added

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedColumnReader.java (hunk 11 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedColumnReader.java]: ok

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java (hunk 12 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java]: ok

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/position/RowPosition.java (hunk 13 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/position/RowPosition.java]: ok

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedPrimitiveColumnReader.java (hunk 14 above)
Comment (Stephen0421) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/NestedPrimitiveColumnReader.java]: ok

**Comment by Stephen0421** [comment]
> @Stephen0421, could you, please, look at this issue https://issues.apache.org/jira/projects/FLINK/issues/FLINK-35698 It seems to be connected to your work. There is a problem Parquet reading nested data when field type is DECIMAL. I provided a test case there.

Hello @empathy87, I run your test case in my local pr branch and it run successfully, I think the pr has solve the problem.

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**